Do not attempt to find an executable if a script is referenced

adinapoli opened the pull request:
@andreasabel This is a (potentially naive) attempt at fixing #189 (and perhaps #107?).

This was the result of a quick hacking session so I am sure the story is more complicated than this, and perhaps we should also have @gregwebs chime in on this one.

~~I am not sure _why_ the switch from `system-filepath` to `filepath` triggered it~~, but the crux of the issue is that unless `shelly` is run with `escaping False`, the input `FilePath` fed into `run` & co will be "escaped", which in this context means calling the `runCommand` function, which will attempt to find an executable at `PATH`.

This works most of the time, but if we have something like `./foo/bar/baz.sh`, this won't work, and `Shelly` will complain.

The easiest fix I could think of was to simply check if we had a `FilePath` commencing with `.` and, in this case, we don't attempt to find an executable, but we run the command as-if we used `escaping False`.

I have also added a regression test, to catch this in the future.

Thoughts?

**EDIT:** For visibility, here is the postmortem copied from an inline comment:

 I have finally nailed down exactly why the regression happened. We have this code, which is responsible for locating the file, in the guts of `whichEith`:

```hs
    whichFull fp = do
      (trace . mappend "which " . toTextIgnore) fp >> whichUntraced
      where
        whichUntraced | isAbsolute fp          = checkFile
                      | dotSlash splitOnDirs   = checkFile
                      | length splitOnDirs > 0 = lookupPath  >>= leftPathError
                      | otherwise              = lookupCache >>= leftPathError

        splitOnDirs = splitDirectories fp
        dotSlash ("./":_) = True
        dotSlash _ = False
```

As you can see, we are already accounting for `dotSlash`, so why things fail? Well, it turns out that both `system-filepath`  and `filepath` have a `splitDirectory` function, but hey behave differently:

```
> import qualified Filesystem.Path.CurrentOS as SFP
> import qualified Data.Text as T
> import System.FilePath as FP
> FP.splitDirectories "./test/data/hello.sh"
[".","test","data","hello.sh"]
> SFP.splitDirectories (SFP.fromText $ T.pack "./test/data/hello.sh")
[FilePath "./",FilePath "test/",FilePath "data/",FilePath "hello.sh"]
> 
```

As you can see the former doesn't include the "./", so the pattern match on `dotSlash` fails, and here is our regression.

adinapoli commented:
In a lucky turn of event I will be working remotely from Rome and my digital nomad workstation is a Linux machine (as opposed to my main Mac mini), so I can definitely look into this one later this week.

Thanks for the heads up!

gregwebs commented:
Thanks for adding a test case. With enough testing we can get make a change.

andreasabel commented on src/Shelly.hs:
```diff
@@ -325,22 +325,26 @@ runCommand handles st exe args = findExe exe >>= \fullExe ->
-          -- it would be better to specifically detect that bug
+      = case fp of
+          -- If the 'FilePath' contains a more articulated path than just
+          -- an executable, don't try to find it.
```
Should this rather check whether there is a directory component in `exe`? What about `bla/hello.sh` instead of `./hello.sh`?  Currently you would accept e.g. `.bla/hello.sh` or `.hello.sh`.
I think checking for _starting with dot_ isn't the correct criterion.

adinapoli replied:
Yes, I take your point this is too simplistic. I will have to think about this a little bit.

adinapoli replied:
@andreasabel Aha! After a more thorough look, I have finally nailed down exactly why the regression happened. We have this code, which is responsible for locating the file, in the guts of `whichEith`:

```hs
    whichFull fp = do
      (trace . mappend "which " . toTextIgnore) fp >> whichUntraced
      where
        whichUntraced | isAbsolute fp          = checkFile
                      | dotSlash splitOnDirs   = checkFile
                      | length splitOnDirs > 0 = lookupPath  >>= leftPathError
                      | otherwise              = lookupCache >>= leftPathError

        splitOnDirs = splitDirectories fp
        dotSlash ("./":_) = True
        dotSlash _ = False
```

As you can see, we are already accounting for `dotSlash`, so why things fail? Well, it turns out that both `system-filepath`  and `filepath` have a `splitDirectory` function, but hey behave differently:

```
> import qualified Filesystem.Path.CurrentOS as SFP
> import qualified Data.Text as T
> import System.FilePath as FP
> FP.splitDirectories "./test/data/hello.sh"
[".","test","data","hello.sh"]
> SFP.splitDirectories (SFP.fromText $ T.pack "./test/data/hello.sh")
[FilePath "./",FilePath "test/",FilePath "data/",FilePath "hello.sh"]
> 
```

As you can see the former doesn't include the "./", so the pattern match on `dotSlash` fails, and here is our regression.

andreasabel replied:
Excellent analysis! Thanks, @adinapoli!

andreasabel commented on test/data/hello.sh:
```diff
@@ -0,0 +1,3 @@
+#!/usr/bin/env bash
```
Could this be OS agnostic?
Windows also has scripts (and `echo` is also a Windows command`).

Or, there could be another script (like `hello.bat`) that is then tested under Windows.

adinapoli replied:
I will have to think about this. I think that a `hello.bat` file makes sense, although I am not sure if windows support dots like in unix. Something I have to point out is that (I think?) currently Shelly cannot execute things without the dot unless you use `escaping False`, and this is independent from the `system-filepath` regression. To be precise, I think something like this fails in Shelly:

```hs
      it "script (without dot) at $PWD" $ do
        res <- shelly $ do
          run_ "chmod" ["+x", "test/data/hello.sh"]
          run "test/data/hello.sh" []
        res @?= "Hello!\n"
```

I think that's because that `whichEith` function will consider this _not_ an absolute path (it's relative) and it doesn't have dot in it, so it will try to find the whole thing at `PATH` and things will fail. But I will suggest not trying to patch this in this PR, to not inflate its size.

andreasabel replied:
Ok, we do not have to improve on Shelly in this PR, just restore its old behavior...

andreasabel commented on src/Shelly.hs:
```diff
@@ -605,14 +605,34 @@ whichEith originalFp = whichFull
-                      | otherwise              = lookupCache >>= leftPathError
+        whichUntraced | isAbsolute fp             = checkFile
+                      | startsWithDot splitOnDirs = checkFile
+                      | length splitOnDirs > 0    = lookupPath  >>= leftPathError
```
I don't understand when the guard `length splitOnDirs > 0` can fail (be `False`)... It seems this happens only when `fp` is `[]`.  But this should not happen, should it?  So is the last case dead?
Here are some experiments:
```
ghci> :t splitDirectories
splitDirectories :: FilePath -> [FilePath]
ghci> splitDirectories "."
["."]
ghci> splitDirectories ""
[]
ghci> splitDirectories "/"
["/"]
ghci> splitDirectories "//"
["//"]
ghci> splitDirectories "///"
["///"]
ghci> splitDirectories "/a"
["/","a"]
ghci> splitDirectories "/a/"
["/","a"]
ghci> splitDirectories "/a/b"
["/","a","b"]
ghci> splitDirectories "c:/a"
["c:","a"]
ghci> splitDirectories "c:a"
["c:a"]
ghci> splitDirectories "c:foo.bat"
["c:foo.bat"]
ghci> splitDirectories "c:\\foo.bat"
["c:\\foo.bat"]
ghci> splitDirectories "c:\\\\foo.bat"
["c:\\\\foo.bat"]
ghci> splitDirectories "c:/foo.bat"
["c:","foo.bat"]
```

adinapoli replied:
Which output of `splitDirectories` am I looking at? the one from `system-filepath` or `filepath`? :sweat_smile:

andreasabel replied:
Oh sorry, this is the one of `filepath` which is used in latest `shelly`.  But the `system-filepath` one is not much different.
@gregwebs : Can you explain the mystery of the `length splitOnDirs > 0` guard?  You introduced it in https://github.com/gregwebs/Shelly.hs/commit/d4415838b673fae377481dd73a8de1e60f3b56db#r70391227

andreasabel replied:
Greg agrees that the guard `length splitOnDirs > 0` is unnecessary (and the clause below can be dropped).

adinapoli replied:
Excellent! I will amend this PR with this and the other changes between today and tomorrow. Thanks!

andreasabel commented on src/Shelly.hs:
```diff
@@ -605,14 +605,34 @@ whichEith originalFp = whichFull
+        -- function, but it returns \"./\" as its first argument,
+        -- so we pattern match on both for backward-compatibility.
+        startsWithDot (".":_)  = True
+        startsWithDot ("./":_) = True
```
Does it actually make sense to keep the old test for backwards-compatibility?  I mean, `System.FilePath.splitDirectories`will never return list that contains `"./"`.

adinapoli replied:
It does look harmless -- the only reason why I have kept it was mostly for documenting purposes and to defend ourselves from draconian cases where people would attempt frankestein builds that uses newer Shelly-s but `system-filepath` (assuming that is possible). I suppose it gives us a bit more flexibility at a very little cost, but I am also happy to get rid of the redundant test :wink:

andreasabel replied:
Don't get me wrong: I am very much for documenting the why and also the history.
However, an impossible case shouldn't be there, because it will create insecurity in the reader of the code.  It will actually suggest that the case is possible and should be handled.  If you want to keep it in, it should be
```haskell
startsWithDot ("./":_) = undefined -- this case is impossible
```
but I think this isn't necessary.
I'd keep references to previous behavior in the comment.

adinapoli commented:
> Ok, we do not have to improve on Shelly in this PR, just restore its old behavior...

@andreasabel hehe yes, my temptation would be to eschew any temptation to improve Shelly as part of this PR and instead deliver the smallest fix that restores the old behaviour, and perhaps we flag all these low hanging fruits in a ticket, as those sounds very inviting for newcomers willing to contribute to the project :wink:

adinapoli commented:
@andreasabel Ok, I have addressed all the review remarks. Reshuffling that `whichFull` function also had the added bonus of removing some now-redundant functions.

Something I should point out though is that, at least on my Linux machine, `splitDirectories` **can** return a length of 0, which is when the empty string is passed:

```hs
> import System.FilePath as FP
> FP.splitDirectories ""
[]
> 
```

This is very weird, because you also ran the same test [here](https://github.com/gregwebs/Shelly.hs/pull/216#discussion_r841283815) but got something different. Having said that, the tests seem to still all pass, so perhaps that's nothing to worry about.

Furthermore, I couldn't conceivably see why somebody would try to run a command by passing the empty string as the executable (and I agree with your analysis, garbage in, garbage out). A case could be made that Shelly could get rid of these unrepresentable states if it had its own (internal) definition of something like `newtype FilePath = FilePath (NonEmpty Char)` but again, this might be overkill (or too simplistic) and definitely doesn't belong to this PR :wink:

adinapoli commented:
@andreasabel I have lost track of this -- are you folks expecting anything else from me or is this essentially good to go? Thanks! 😉

andreasabel reviewed (approved):
Thanks again @adinapoli and apologies for the long break.
I am planning to release this as 1.11.0 now, assuming you haven't experienced any problems with this code in the last 9 month.
I rebased this on latest `master` but otherwise left the commits untouched.

adinapoli commented:
> Thanks again @adinapoli and apologies for the long break. I am planning to release this as 1.11.0 now, assuming you haven't experienced any problems with this code in the last 9 month. I rebased this on latest `master` but otherwise left the commits untouched.

No worries, and thanks for picking this one up 😉 In interest of full transparency, I ended up not using (just yet) this branch/commit, as I try to avoid depending on unreleased/unapproved forks/branches/patches for production code, so I just ended up downgrading to a suitable Shelly, but given how much we spent thinking about this issue, I think it should be fine. We know precisely where and why the regression happened and we have a test to prevent that from happening again, so we should be good 😉

andreasabel commented:
Thanks again, @adinapoli !  
I pushed the button now, so you have opportunities to evaluate the fix in production now:  https://hackage.haskell.org/package/shelly-1.11.0

adinapoli commented:
> Thanks again, @adinapoli ! I pushed the button now, so you have opportunities to evaluate the fix in production now: https://hackage.haskell.org/package/shelly-1.11.0

Beautiful, I'll keep you posted, thanks!
